File: java/data_strcture/BinarySearch.java

import java.util.Scanner;

class binarySearch {

    /**
     * @brief in this function print the array elements
     * @param array konte array ko print karana hai
     * @param size array ki size kitani hogi
     */
    static void printArrayElements(int[] array, int size) {
        for (int index = 0; index < size; index++) {
            System.out.println(+array[index]);
        }

    }

    /**
     * @brief in this function sort the array elements
     * @param array konte array ko sort karana hai
     * @param size array ki size kitani hogi
     */
    static void sortArrayElements(int[] array, int size) {
        for (int counter = 0; counter < size; counter++) {
            for (int index = 0; index < (size - 1); index++) {
                if (array[index] > array[index + 1]) {
                    array[index] = array[index] + array[index + 1];
                    array[index + 1] = array[index] - array[index + 1];
                    array[index] = array[index] - array[index + 1];
                }

            }
        }

    }

    /**
     * *
     * @brief in this function search a element from a sorting array
     * @param array konte array mai se element ko search karana hai
     * @param size array size kitani hogi
     * @param key konte element ko search karna hai
     */
    static int binarySearch(int[] array, int size, int key) {
        int base_index = 0;
        int higher_index = size - 1;
        while (base_index <= higher_index) {
            int middle_index = (base_index + higher_index) / 2;
            int middle_key = array[middle_index];
            if (key == middle_key) {
                return middle_index;
            } else if (key < middle_key) {
                higher_index = middle_index - 1;
            } else {
                base_index = middle_index + 1;
            }
        }
        return -1;

    }

    public static void main(String args[]) {
        int size = 6;
        int[] array = {34, 76, 54, 74, 90, 6};
        System.out.println("choice the option");
        System.out.println("1. print the array elements");
        System.out.println("2. sort the array elements");
        System.out.println("3. search the element from array");
        Scanner sc = new Scanner(System.in);
        int choice = sc.nextInt();

        switch (choice) {
            case 1:
                printArrayElements(array, size);
                break;
            case 2:
                sortArrayElements(array, size);
                printArrayElements(array,size);
                break;
            case 3:
                sortArrayElements(array, size);
                printArrayElements(array, size);
                Scanner sc1 = new Scanner(System.in);
                System.out.println("Enter the element to search:");
                int key = sc1.nextInt();
                sc1.close();
                int search_index = binarySearch(array, size, key);
                if(search_index == -1){
                    System.out.println("element is a not found");
                }
                else{
                    System.out.println("The element is found at index: " + search_index);
                     break;
                }
                
            default:
                System.out.println("Invalid choice");
                break;
        }

    }

}

File: data-structures-3sk03/notes/mid_semter_eaxm_question_collactivate/binary_search.c
#include <stdio.h>
/**
 * @brief this function sort the array elements
 * @param array konte array elements ko sort karana hai
 * @param size size of the array
 */
int sort_array_elements(int array[], int size)
{
    for (int counter = 0; counter < (size); counter++)
    {
        for (int index = 0; index < (size - 1); index++)
        {
            if (array[index] > array[index + 1])
            {
                array[index] = array[index] + array[index + 1];
                array[index + 1] = array[index] - array[index + 1];
                array[index] = array[index] - array[index + 1];
            }
        }
    }
}

/**
 * @brief this function print the array elements on std screen
 * @param array konte array elements ko print karana hai
 * @param size array ki size kitani rahegi
 */
int print_array_elements(int array[], int size)
{
    for (int index = 0; index < (size); index++)
    {
        printf("%i\n", array[index]);
    }
}

/**
 * @brief this  function search element from the sort array
 * @param array konte array mai se element search karana hai
 * @param size array ki size kitani hai
 * @param key konte element ko search karana hai
 * @return int
 */
int binary_search(int array[], int size, int key)
{
    int base_index = 0;
    int higher_index = size - 1;
    while (base_index <= higher_index)
    {
        int middle_index = (base_index + higher_index) / 2;
        int middle_key = array[middle_index];

        if (middle_key == key)
        {
            printf("%i\n", middle_index);
            return middle_index;
        }
        else if (key < middle_key)
        {

            higher_index = middle_index - 1;
        }
        else
        {
            base_index = middle_index + 1;
        }
    }
    return -1;
}

int main(void)
{
    int size = 6;
    int array[] = {43, 100, 89, 56, 34, 70};
    int choice;
    printf("enter 1 to print the array element:\n");
    printf("enter 2 sort array elements:\n");
    printf("enter 3 for binary search element:\n");
    printf("enter a choice:\n");
    scanf("%i", &choice);
    switch (choice)
    {
    case 1:
        print_array_elements(array, size);
        break;
    case 2:
        sort_array_elements(array, size);
        print_array_elements(array, size);
        break;
    case 3:
        int key_element;
        printf("please choice the key from a array to find:\n");
        sort_array_elements(array, size);
        print_array_elements(array, size);
        printf("enter a element:");
        scanf("%i", &key_element);
        printf("element find at a index:");
        binary_search(array, size, key_element);
        break;
    default:
        break;
    }
    return 0;
}
